### ISSUE TITLE
Add submitted job IDs to a log file

### ISSUE THREAD

**Issue by FelixKrueger** [ewels/clusterflow#31]
Hi Phil, we recently had issues with CF runs failing for various reasons, probably mostly memory related issues. Once the jobs were killed there is no way of finding out what the job ID was and this made it almost impossible to find the folder (difficult to track when hundreds of single-cell RNA-Seqs are running at the same time...) or get information of what happened to the job using `qacct -j <ID>`. Something like this would be helpful:

Your job 259132 ("cf_bismark_1423231738_bismark_align_295") has been submitted
Your job 259133 ("cf_bismark_1423231738_bismark_deduplicate_005") has been submitted
Mmmkay?

**Comment by ewels** [comment]
Yes, but typically the job-submitting is done by the main `cf` executable in interactive mode, so this information would just be spat out to `STDERR` and not stored in a log file anywhere.

Secondly, to get the submitted job ID values (not just the job name) is a bit messy, as this will be different for every queue system. SLURM actually already prints this info to `STDOUT` when the jobs are submitted, so no change is needed there.

If you're running the dev version you can just use the `--verbose` option and it will print the full job submission commands (which include the job ID within them) to `STDERR`. Is this enough for you?

Is there really no way to find the IDs or names of killed jobs? That seems like the kind of thing that would end up in a cluster log file somewhere.. Surely they don't just get killed silently?

**Comment by ewels** [comment]
Would [this](http://stackoverflow.com/questions/16570895/how-to-get-a-list-of-failed-job-from-sge) help?

**Comment by FelixKrueger** [comment]
Can't you just print the text we see on screen to a Clusterflow log? That would already help... E.g.:

Your job 259132 ("cf_bismark_1423231738_bismark_align_295") has been submitted
Your job 259133 ("cf_bismark_1423231738_bismark_deduplicate_005") has been submitted
Your job 259134 ("cf_bismark_1423231738_bismark_methXtract_600") has been submitted
Your job 259135 ("cf_bismark_1423231738_bismark_report_584") has been submitted
Your job 259136 ("cf_bismark_1423231738_email_run_complete_610") has been submitted
Your job 259137 ("cf_bismark_1423231738_bismark_align_839") has been submitted
Your job 259138 ("cf_bismark_1423231738_bismark_deduplicate_199") has been submitted
Your job 259139 ("cf_bismark_1423231738_bismark_methXtract_923") has been submitted
Your job 259140 ("cf_bismark_1423231738_bismark_report_276") has been submitted
Your job 259141 ("cf_bismark_1423231738_email_run_complete_874") has been submitted
Your job 259142 ("cf_bismark_1423231738_bismark_align_689") has been submitted
Your job 259143 ("cf_bismark_1423231738_bismark_deduplicate_764") has been submitted
Your job 259144 ("cf_bismark_1423231738_bismark_methXtract_516") has been submitted
Your job 259145 ("cf_bismark_1423231738_bismark_report_558") has been submitted
Your job 259146 ("cf_bismark_1423231738_email_run_complete_590") has been submitted
Your job 259147 ("cf_bismark_1423231738_bismark_align_215") has been submitted
Your job 259148 ("cf_bismark_1423231738_bismark_deduplicate_716") has been submitted
Your job 259149 ("cf_bismark_1423231738_bismark_methXtract_881") has been submitted
Your job 259150 ("cf_bismark_1423231738_bismark_report_862") has been submitted
Your job 259151 ("cf_bismark_1423231738_email_run_complete_664") has been submitted
Your job 259152 ("cf_bismark_1423231738_bismark_align_832") has been submitted
Your job 259153 ("cf_bismark_1423231738_bismark_deduplicate_550") has been submitted
Your job 259154 ("cf_bismark_1423231738_bismark_methXtract_819") has been submitted
Your job 259155 ("cf_bismark_1423231738_bismark_report_579") has been submitted
Your job 259156 ("cf_bismark_1423231738_email_run_complete_847") has been submitted
Your job 259157 ("cf_bismark_1423231738_bismark_align_284") has been submitted
Your job 259158 ("cf_bismark_1423231738_bismark_deduplicate_845") has been submitted
Your job 259159 ("cf_bismark_1423231738_bismark_methXtract_835") has been submitted
Your job 259160 ("cf_bismark_1423231738_bismark_report_444") has been submitted
Your job 259161 ("cf_bismark_1423231738_email_run_complete_897") has been submitted
Your job 259162 ("cf_bismark_1423231738_bismark_align_115") has been submitted
Your job 259163 ("cf_bismark_1423231738_bismark_deduplicate_398") has been submitted
Your job 259164 ("cf_bismark_1423231738_bismark_methXtract_250") has been submitted
Your job 259165 ("cf_bismark_1423231738_bismark_report_615") has been submitted
Your job 259166 ("cf_bismark_1423231738_email_run_complete_006") has been submitted
Your job 259167 ("cf_bismark_1423231738_bismark_align_451") has been submitted
Your job 259168 ("cf_bismark_1423231738_bismark_deduplicate_237") has been submitted
Your job 259169 ("cf_bismark_1423231738_bismark_methXtract_386") has been submitted
Your job 259170 ("cf_bismark_1423231738_bismark_report_180") has been submitted
Your job 259171 ("cf_bismark_1423231738_email_run_complete_745") has been submitted
Your job 259172 ("cf_bismark_1423231738_bismark_align_830") has been submitted
Your job 259173 ("cf_bismark_1423231738_bismark_deduplicate_326") has been submitted
Your job 259174 ("cf_bismark_1423231738_bismark_methXtract_739") has been submitted
Your job 259175 ("cf_bismark_1423231738_bismark_report_908") has been submitted
Your job 259176 ("cf_bismark_1423231738_email_run_complete_033") has been submitted
Your job 259177 ("cf_bismark_1423231738_bismark_align_612") has been submitted
Your job 259178 ("cf_bismark_1423231738_bismark_deduplicate_680") has been submitted
Your job 259179 ("cf_bismark_1423231738_bismark_methXtract_195") has been submitted
Your job 259180 ("cf_bismark_1423231738_bismark_report_416") has been submitted
Your job 259181 ("cf_bismark_1423231738_email_run_complete_031") has been submitted
Your job 259182 ("cf_bismark_1423231738_bismark_align_079") has been submitted
Your job 259183 ("cf_bismark_1423231738_bismark_deduplicate_098") has been submitted
Your job 259184 ("cf_bismark_1423231738_bismark_methXtract_188") has been submitted
Your job 259185 ("cf_bismark_1423231738_bismark_report_385") has been submitted
Your job 259186 ("cf_bismark_1423231738_email_run_complete_340") has been submitted
Your job 259187 ("cf_bismark_1423231738_bismark_align_861") has been submitted
Your job 259188 ("cf_bismark_1423231738_bismark_deduplicate_028") has been submitted
Your job 259189 ("cf_bismark_1423231738_bismark_methXtract_332") has been submitted
Your job 259190 ("cf_bismark_1423231738_bismark_report_286") has been submitted
Your job 259191 ("cf_bismark_1423231738_email_run_complete_290") has been submitted
Your job 259192 ("cf_bismark_1423231738_email_pipeline_complete_081") has been submitted

**Comment by ewels** [comment]
I'm not sure what those status messages that you're posting are - they don't come from Cluster Flow (I don't think). Are you sure that they aren't being printed by SGE itself?

I guess I could try to capture this by piping `STDERR` and `STDOUT` to a new clusterflow command log file when it's fired. Hmm... Would need to have a play to figure out how to actually make that work.

**Comment by FelixKrueger** [comment]
Great, that's what I wanted to hear. Have a good weekend.

**Comment by ewels** [comment]
Yeah, actually this is just what I'm doing at the moment with the SLURM output actually, though I'm not writing it to file. I guess I could create a new 'submission_log.txt' or something and print captured text to it.

**Comment by ewels** [comment]
Ok, all done @FelixKrueger - I can't test it on GRID Engine, but works nicely on SLURM. Let me know..

**Comment by FelixKrueger** [comment]
Thanks, will do!

**Comment by ewels** [comment]
@FelixKrueger - I also just added a little more, job submission commands are also appended to the end of the log file. Hope that helps.

**Comment by ewels** [comment]
@stu2 - this new functionality kind of duplicates and adds to what `--verbose` does. Would you mind if I remove `--verbose` for simplicity's sake, or would you rather I keep it?

**Comment by stu2** [comment]
Hi Phil,
Yes that's fine with me, --verbose isn't really a core part of the pipeline.

**Comment by ewels** [comment]
@stu2 - along a similar line, I really like your idea about printing software versions to logs. Now I'm thinking about removing `--print_versions` and just getting all modules to _always_ print the software versions to the log when they run.

I can't think of any disadvantage to having this there, other than sometimes making the log file a bit long and unwieldy (could maybe try to clean some of the longer outputs by searching for version number). What do you think?

**Comment by stu2** [comment]
Yeah, it's not elegant but its better to have the info in there than not. A parser to filter the version numbers from the help screens past, present and future would be useful, but tricky to code!

**Comment by ewels** [comment]
So I've just gone through all of the modules updating this. They now all print to `STDERR` after the module headers are printed, so all should be in the log files.

I had a look at each version call as I went, most were actually fairly concise. For the longer ones, I just added a `2>&1 | head -n 5` (or something similar) to the call, instead of actual parsing. In each module that I did this with, I loaded the oldest version that we have in our modules system to check that it still worked.
